crypto.go

```diff
@@ -34,37 +43,68 @@ func Equals(sk, other []byte) bool {
 
 // Verify checks the given signature and returns true if it is valid.
 func Verify(pk, msg, signature []byte) bool {
+	pub, err := secpEc.NewPublicKey(pk)
+	if err != nil {
+		return false
+	}
+
+	ecsdaOpts := &secpEc.ECDSAOptions{Encoding: secpEc.EncodingCompact, RejectMalleable: true, Hash: crypto.SHA256}
+
+	// Check if the signature is in [R | S | V] style.
+	// if so, we change the Encoding to "EncodingCompactRecoverable"
 	if len(signature) == 65 {
-		// Drop the V (1byte) in [R | S | V] style signatures.
+		// Normally you would drop the V (1byte) in [R | S | V] style signatures.
 		// The V (1byte) is the recovery bit and is not apart of the signature verification.
-		return secp256k1.VerifySignature(pk[:], msg, signature[:len(signature)-1])
+		// EncodingCompactRecoverable expects the V to be included in the signature
+		ecsdaOpts.Encoding = secpEc.EncodingCompactRecoverable
 	}
 
-	return secp256k1.VerifySignature(pk[:], msg, signature)
+	return pub.Verify(msg, signature, ecsdaOpts)
 }
 
 // GenerateKeyFromSeed generates a new key from the given reader.
-func GenerateKeyFromSeed(seed io.Reader) ([]byte, error) {
-	key, err := ecdsa.GenerateKey(secp256k1.S256(), seed)
-	if err != nil {
-		return nil, err
-	}
-
-	privkey := make([]byte, PrivateKeyBytes)
-	blob := key.D.Bytes()
-
-	// the length is guaranteed to be fixed, given the serialization rules for secp2561k curve points.
-	copy(privkey[PrivateKeyBytes-len(blob):], blob)
-
-	return privkey, nil
+// Deprecated: use GenerateKey instead - the given reader is no longer used.
+func GenerateKeyFromSeed(io.Reader) ([]byte, error) {
+	return GenerateKey()
 }
 
 // GenerateKey creates a new key using secure randomness from crypto.rand.
 func GenerateKey() ([]byte, error) {
-	return GenerateKeyFromSeed(rand.Reader)
+	key, err := secpEc.GenerateKey()
+	if err != nil {
+		return nil, err
+	}
+
+	return key.Bytes(), nil
 }
 
 // EcRecover recovers the public key from a message, signature pair.
 func EcRecover(msg, signature []byte) ([]byte, error) {
```

Comment:
We need to assert that `msg` is 32 bytes exactly. The underlying library will truncate to 32 bytes but won't fail if there's garbage data.

Reply:
Note: it will assert on verify/sign, just not on recovery.

Reply:
Ah makes sense.  Please let me know if there are things that I can assist with on my end.

Reply:
Other than not giving anyone by the name "Jia Tan" commit access :laughing:?

But in all seriousness, thanks for writing this library. I think we're all good here, just waiting on another round from @Schwartz10.